### PR TITLE
[WIP] Supports NGSIv2

### PR DESCRIPTION
**Pull request by dcalvoalonso**

### CODE REVIEW

File: lib/commonBindings.js [hunk 1]
```diff
@@ -19,6 +19,8 @@
  *
```
Comment (fgalan) [inline lib/commonBindings.js]: CNR entry missing

File: lib/configService.js [hunk 2]
```diff
@@ -114,7 +116,23 @@ function getLogger() {
     return logger;
 }
 
+/**
+ * It checks if the configuration file states the use of NGSIv3
+ *
+ * @return     {boolean}  Result of the checking
+ */
+function checkNgsi2() {
```
Comment (fgalan) [inline lib/configService.js]: (Similar comment to one in the LWM2M repo PR)

I think to remember the same check at IOTA Library... Do we need double-check (one in the library, one in the agent code)? Could duplicated code be avoided?

File: lib/commonBindings.js (hunk 1 above)
Comment (dcalvoalonso) [reply lib/commonBindings.js]: Fixed in 292d293860f9f23dd860cfb60b0111605df0f7f6

**Comment by dcalvoalonso** [comment]
I have tagged this PR as WIP because the branch relies on a branch of https://github.com/telefonicaid/iotagent-node-lib that is not yet merged (telefonicaid/iotagent-node-lib#556).

Nevertheless, in my eyes, the implementation of the functionalities is completely ready in this IOTA.

**Comment by fgalan** [comment]
Ok. I'll have a look as soon as we can. Thanks!

File: CHANGES_NEXT_RELEASE [hunk 3]
```diff
@@ -3,3 +3,5 @@ Allow get list of commands without sending measures (empty payload) [#256]
 Fix transport in autoprovision device depending on binding [#257]
 Fix: defaultKey in config.js (supposely fixing #222 and #207)
 Fix default resource /iot/json instead of /iot/d in config and tests
+Add: allow NGSIv2 for updating active attributes at CB, through configuration (#250)
+Add: NGSIv2 support for data transformation plugins (#250)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: What "data tranformation plugins" mean in this context? The ones at https://github.com/telefonicaid/iotagent-node-lib#-data-mapping-plugins? Wasn't that adaptation done in the context of the iotagent-node-lib?

Just to understand the content of the CNR... :)

File: lib/iotagent-json.js [hunk 4]
```diff
@@ -19,6 +19,8 @@
  *
  * For those usages not covered by the GNU Affero General Public License
  * please contact with::[contacto@tid.es]
+ *
+ * Modified by: Daniel Calvo - ATOS Research & Innovation
```
Comment (fgalan) [inline lib/iotagent-json.js]: Was this file actually modified?

File: package.json [hunk 5]
```diff
@@ -52,10 +52,10 @@
     "body-parser": "1.15.0",
     "dateformat": "1.0.12",
     "amqplib": "^0.5.1",
-    "iotagent-node-lib": "https://github.com/telefonicaid/iotagent-node-lib.git#master",
+    "iotagent-node-lib": "https://github.com/dcalvoalonso/iotagent-node-lib.git#task/exportConfigNgsi2",
     "logops": "1.0.0-alpha.7",
     "mqtt": "1.7.0",
-    "request": "^2.69.0",
+    "request": "2.39.0 - 2.81.0",
```
Comment (fgalan) [inline package.json]: Is this change really required? We should try to change as little dependencies as possible. I understand that the change in nock may be needed if the mock included in the new tests require so, but what about the change in request?

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+ * please contact with::[contacto@tid.es]
+ *
+ * Modified by: Daniel Calvo - ATOS Research & Innovation
+ */
+
+/* jshint camelcase: false */
+
+'use strict';
+
+var iotagentMqtt = require('../../../'),
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: The name of the variable is a bit missleading, as the test is focused in HTPP. Maybe just `iotagent` would be fine.

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+            }
+        };
+
+        nock.cleanAll();
+
+        contextBrokerMock = nock('http://192.168.1.1:1026')
+            .matchHeader('fiware-service', 'smartGondor')
+            .matchHeader('fiware-servicepath', '/gardens')
+            .post('/v1/updateContext')
+            .reply(200, utils.readExampleFile('./test/contextResponses/multipleMeasuresSuccess.json'));
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: This is a NGSIv1 update response... I understand this is needed to mock the entity creation upon device provision. Is that correct?

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+            iotAgentLib.clearAll,
+            iotagentMqtt.stop
+        ], done);
+    });
+
+    describe('When a POST measure arrives for the HTTP binding', function() {
+        var optionsMeasure = {
+            url: 'http://localhost:' + config.http.port + '/iot/json',
+            method: 'POST',
+            json: {
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: It would be great to include some tests for JSON native types different from string, eg:

```
            json: {
                humidity: 32,
                active: true
            },
```

in order to check if the IOTAgent is able to push them with the correct native type to CB. Note this is a new possiblity with NGSIv2 that the agent hasn't needed to consider before (as with NGSIv1 no matter if the agent was able to push native types, they get "stringfied" by Orion upon reception time).

Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: (Maybe the PR include tests like that but I haven't reach them yet in my review... however, I'm writting the comment now so I don't forget :)

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 7 above)
Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: (General comment for all the `.post('/v1/updateContext')` cases in the PR)

Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: However, not sure if I'm getting this correctly... in test/contextResponses/multipleMeasuresSuccess.json I see:

```
"type": "MQTTSensor",
"id": "MQTTId_2"
```

which doesn't correct to what it should be taking into account the provisioning from test/deviceProvisioning/provisionDeviceHTTP.json

```
"entity_name": "Second MQTT Device",
"entity_type": "AnMQTTDevice",
```

Maybe I'm missing something...

(This problem may affect to other similar cases along this PR)

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+                },
+                qs: {
+                    i: 'dev0130101',
+                    k: '1234'
+                }
+            },
+            provisionOptions = {
+                url: 'http://localhost:' + config.iota.server.port + '/iot/devices',
+                method: 'POST',
+                json: utils.readExampleFile('./test/deviceProvisioning/provisionDeviceTimeinstant.json'),
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Which is the role of the `"expression": "${@humedad / 1000}",` in the JSON file used for provision in this case? It seems to be orthoghonal to the test...

File: test/unit/ngsiv2/contextRequests/timeInstantMeasures.json [hunk 10]
```diff
@@ -0,0 +1,16 @@
+{
+  "humidity":{
+    "type": "percent",
+    "value": "111222",
+    "metadata":{
+      "TimeInstant":{
```
Comment (fgalan) [inline test/unit/ngsiv2/contextRequests/timeInstantMeasures.json]: This (and the other TimeInstance usage in L12 below) could be problematic at CB. Is `20200222T222222` a valid ISO8601 format?

I'd suggest to test this literrarly (e.g. a crafted curl using this very same JSON file) in a real Orion instance.

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+            json: {
+                humidity: '32',
+                temperature: '87'
+            },
+            headers: {
+                'fiware-service': 'smartGondor',
+                'fiware-servicepath': '/gardens'
+            },
+            qs: {
+                i: 'MQTT_UNPROVISIONED',
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: HTTP_UNPROVISIONED ?

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v1/updateContext')
+                .reply(200, utils.readExampleFile('./test/contextResponses/multipleMeasuresSuccess.json'));
+
+            contextBrokerMock
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v2/entities/TheLightType:MQTT_UNPROVISIONED/attrs',
+                 utils.readExampleFile('./test/unit/ngsiv2/contextRequests/unprovisionedDevice.json'))
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: I see a test/contextRequests/unprovisionedDevice.json on the repository, but not a test/unit/**ngsiv2**/contextRequests/unprovisionedDevice.json. Neither in the PR.

Maybe the file was missed to be added to the PR?

Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Never mind... I found it at the end in the PR

NTC, sorry the noise

File: CHANGES_NEXT_RELEASE (hunk 3 above)
Comment (dcalvoalonso) [reply CHANGES_NEXT_RELEASE]: Yes, These are the data transformation plugins. As you can see, there are no real modifications in this PR with respect to the plugins. I just introduced this statement to reflect that NGSIv2 is supported.

File: package.json (hunk 5 above)
Comment (dcalvoalonso) [reply package.json]: This change is require so that unit tests are passed with modern versions of NodeJs. As you can see, the same versions are specified in https://github.com/telefonicaid/iotagent-node-lib/blob/master/package.json#L45. This change was introduced by https://github.com/telefonicaid/iotagent-node-lib/pull/560 to fix unit tests.

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 7 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: You're correct, this part of the tests is related to device provisioning, which as you know has not been migrated to NGSIv1.

You are again right that this is a bit strange because the mock for the context broker should return something like:
```
"type": "AnMQTTDevice",
"id": "Second MQTT Device"
```

Nevertheless, this is not very relevant since the iotagent-node-lib does not check the contents of the answer of the CB. You can check this in https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/devices/deviceService.js#L73. As you can see, it only checks the existence of an answer, with a body and with 200 status code.

For the NGSIv2 tests, we have just `migrated` NGSIv1 tests (i.e. https://github.com/Fiware/iot.IoTagent-JSON/blob/master/test/unit/HTTP_reveice_measures-test.js#L84). I guess that the original developers of this agent knew about this and just decided to simplify the tests, reusing always the same payload for the CB answers to device provisioning requests.

File: test/unit/ngsiv2/amqpBinding-test.js [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+                .matchHeader('fiware-service', 'TestService')
+                .matchHeader('fiware-servicepath', '/testingPath')
+                .post('/v1/updateContext')
+                .reply(200, utils.readExampleFile('./test/contextResponses/multipleMeasuresSuccess.json'));
+
+
+            contextBrokerMock
+                .matchHeader('fiware-service', 'TestService')
+                .matchHeader('fiware-servicepath', '/testingPath')
+                .post('/v2/entities/SensorMachine:MQTT_UNPROVISIONED/attrs',
```
Comment (fgalan) [inline test/unit/ngsiv2/amqpBinding-test.js]: AMQP_UNPROVISIONED ?

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 6 above)
Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Actually, looking to https://github.com/telefonicaid/iotagent-json/pull/265/files#diff-3095e45b7333ce58bd16db6e665000fbR28, `iotaJson` seems to be the perfect name for the variable in all test files :)

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 9 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: I guess that this is an error because there is no `humedad` attribute. Again, we have inherited this from the corresponding NGSIv1 version. Probably they intended to test the expression plugin and for some reason... they changed their mind! :)

File: lib/iotagent-json.js (hunk 4 above)
Comment (dcalvoalonso) [reply lib/iotagent-json.js]: Fixed in 23824ae065ac9129fc666cbbe3dffa8d62678ac1

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 6 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed in 23824ae065ac9129fc666cbbe3dffa8d62678ac1

Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Just to add that the same applies to current NGSIv1 unit tests: https://github.com/Fiware/iot.IoTagent-JSON/blob/master/test/unit/HTTP_reveice_measures-test.js#L28

File: test/unit/ngsiv2/amqpBinding-test.js [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+ *
+ * For those usages not covered by the GNU Affero General Public License
+ * please contact with::[iot_support@tid.es]
+ *
+ * Modified by: Daniel Calvo - ATOS Research & Innovation
+ */
+
+'use strict';
+
+var iotagentMqtt = require('../../../'),
```
[the remark this answers is not in the document]
Comment (dcalvoalonso) [reply test/unit/ngsiv2/amqpBinding-test.js]: Same applies to NGSIv1 test: https://github.com/telefonicaid/iotagent-json/blob/master/test/unit/amqpBinding-test.js

File: lib/configService.js (hunk 2 above)
Comment (fgalan) [reply lib/configService.js]: Fixed in ae72f520f66e348c280580e8baca8306f93ffd76 (it seems so)

File: test/unit/ngsiv2/amqpBinding-test.js (hunk 13 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/amqpBinding-test.js]: Functionally, MQTT_UNPROVISIONED is correct because as you can see in https://github.com/dcalvoalonso/iotagent-json/blob/task/activeAttsNgsi2/test/unit/ngsiv2/amqpBinding-test.js#L149, the same name is used to publish send the update using AMQP.

Again, we have used as baseline the corresponding NGSIv1 and I imagine that this comes from old versions of the IOTA. Maybe, a correct term would be JSON_UNPROVISIONED?

Again, to be coherent, NGSIv1 should be also updated.

File: CHANGES_NEXT_RELEASE (hunk 3 above)
Comment (fgalan) [reply CHANGES_NEXT_RELEASE]: I understand... Whata about merging the two lines into one saying the following?

```
Add: allow NGSIv2 for updating active attributes at CB, through configuration based on iotagent-node-lib (#250)
```

That way it makes clear that it isn't any real modifications in the IOTAgent code itself with respect to the plugins.

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 11 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: As it happens with https://github.com/telefonicaid/iotagent-json/pull/265#pullrequestreview-90500209, I understand that this refers to the IOTA and no to the transport binding. I would rather use JSON_UNPROVISIONED.

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 6 above)
Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: With regards to NGSIv1, probably it doesn't woth the time to fix them. But at least let's ensure that new NGSIv2 unit test are as polished as we can :)

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 7 above)
Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: If the response to /v1/updateContext is not relevant, I'd suggest to simplify to replace the pattern:

```
.reply(200, utils.readExampleFile('./test/contextResponses/multipleMeasuresSuccess.json'));
```

with

```
// Note /v1/updateContext response is not processed by IOTA so its content is irrelevant, as fara as it is a 200 OK
.reply(200, {});
```

This would reduce entropy (and confussions like mine :).

File: CHANGES_NEXT_RELEASE (hunk 3 above)
Comment (dcalvoalonso) [reply CHANGES_NEXT_RELEASE]: Fixed in e95216c36df695c6d4a2a097d4bc8cace20b8625

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 9 above)
Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Prio to this PR is seems that .json file is used only from a single test file NGSIv1 (https://github.com/telefonicaid/iotagent-json/blob/master/test/unit/HTTP_reveice_measures-test.js). Thus, I'd suggest to remove the `"expression": "${@humedad / 1000}"` from it. It is a simple modification that would works for both NGSIv2 and NGSIv1 (travis will ensure ;) and fix the JSON file.

File: test/unit/ngsiv2/amqpBinding-test.js (hunk 13 above)
Comment (fgalan) [reply test/unit/ngsiv2/amqpBinding-test.js]: I like JSON_UNPROVISIONED.

With regards to NGSIv1, I'll let them as they are now. If NGSIv1 code is transient (although the transitory period can be long ;) probably doesn't worth the time to fix also there.

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 7 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: This would not work because the iotagent-node-lib does check if the CB returns body... The only option to change this and to make answers coherent would be to replace this by the appropriate answers from the CB...

Comment (fgalan) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: `{}` is a valid response body? (valid in the sense iotagent-node-lib will not complaint about it)

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 9 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed in e2760b3aee7d517d554809fed0e2f7f1a2102a13

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 11 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed in e2760b3aee7d517d554809fed0e2f7f1a2102a13

File: test/unit/ngsiv2/amqpBinding-test.js (hunk 13 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/amqpBinding-test.js]: Fixed in e2760b3aee7d517d554809fed0e2f7f1a2102a13

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 7 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: It seems to work! :+1:

Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed in 3dbf32d1862f7f3ca4d1658b76c6dccdeaed3874

File: package.json (hunk 5 above)
Comment (fgalan) [reply package.json]: Ok. Makes sense. NTC

File: test/unit/ngsiv2/contextRequests/timeInstantMeasures.json (hunk 10 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/contextRequests/timeInstantMeasures.json]: You are right. Orion anwers:

```
{"error":"BadRequest","description":"date has invalid format"}
```

Comment (fgalan) [reply test/unit/ngsiv2/contextRequests/timeInstantMeasures.json]: Now we have the result of the test (thx!) next question is:

Is `20200222T222222` a valid ISO8601 (but not supported by Orion)? Or an invalid ISO8601 string at all?

Comment (dcalvoalonso) [reply test/unit/ngsiv2/contextRequests/timeInstantMeasures.json]: I have just tried https://www.npmjs.com/package/iso8601-validator and it recognizes `20200222T222222` as a valid ISO8601 string.

I have also checked this in some other resources and it seems that is format is what they call ``basic format`:
- From https://en.wikipedia.org/wiki/ISO_8601:
> For example, "5 April 1981" may be represented as either "1981-04-05"[12] in the extended format or "19810405" in the basic format.

> So a time might appear as either "134730" in the basic format or "13:47:30" in the extended format.

Comment (dcalvoalonso) [reply test/unit/ngsiv2/contextRequests/timeInstantMeasures.json]: My conclusion: 20200222T222222 is a valid ISO8601 string using basic format, which seems to be not supported by Orion.

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 8 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed with ccbe539170af3427946f75901b6f8e5540f158ad

File: test/unit/ngsiv2/contextRequests/timeInstantMeasures.json (hunk 10 above)
Comment (fgalan) [reply test/unit/ngsiv2/contextRequests/timeInstantMeasures.json]: Your analysis seems to be correct. The following documentation describes the ISO8601 support in Orion for DateType: https://fiware-orion.readthedocs.io/en/master/user/ngsiv2_implementation_notes/index.html#datetime-support and basic type part of that support.

I have created an issue about the new Orion functionality: https://github.com/telefonicaid/fiware-orion/issues/3085. In the hope that can come sooner than latter, let's keep the PR as it is now (i.e. mock working as if Orion would have implemented https://github.com/telefonicaid/fiware-orion/issues/3085) and take this into account in the future (e.g. to add a note to IOTAgent node lib documentation explaining that data compression plugin is not going to work with v2 configuration).

NTC (with regards the present PR), from my point of view.

**Comment by fgalan** [comment]
@dcalvoalonso , as far as I see, all the comment threads have been addressed and ended with its correspoding Fixme/NTC mark. Should I re-review now? Or do you plan additional commits? Travis status is fail so I wonder if you are yet working in fixing it... (in which case I'd wait for the final version of the PR to review)

**Comment by dcalvoalonso** [comment]
@fgalan, in my eyes, the PR is completed and I don't plan to submit more commits. You can re-review!! :)

File: package.json [hunk 15]
```diff
@@ -47,16 +47,17 @@
...
+    "amqplib": "^0.5.1",
     "async": "1.5.2",
-    "express": "^4.11.2",
     "body-parser": "1.15.0",
     "dateformat": "1.0.12",
-    "amqplib": "^0.5.1",
-    "iotagent-node-lib": "https://github.com/telefonicaid/iotagent-node-lib.git#master",
+    "express": "^4.11.2",
+    "iotagent-node-lib": "https://github.com/dcalvoalonso/iotagent-node-lib.git#task/exportConfigNgsi2",
+    "iso8601-validator": "^1.0.1",
```
Comment (fgalan) [inline package.json]: Is this actually used? Or it was added only for the test you mention at https://github.com/telefonicaid/iotagent-json/pull/265/files#r163170556 and should be removed?

File: test/unit/ngsiv2/MQTT_receive_measures-test.js [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+                done();
+            });
+        });
+        it('should send its value to the Context Broker', function(done) {
+            var values = {
+                humidity: '32',
+                temperature: '87'
+            };
+
+            mqttClient.publish('/KL223HHV8732SFL1/MQTT_UNPROVISIONED/attrs', JSON.stringify(values), null,
```
Comment (fgalan) [inline test/unit/ngsiv2/MQTT_receive_measures-test.js]: JSON_UNPROVISIONED?

File: test/unit/ngsiv2/MQTT_receive_measures-test.js [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+            contextBrokerMock
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v1/updateContext')
+                .reply(200, '{}');
+
+            contextBrokerMock
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v2/entities/TheLightType:MQTT_UNPROVISIONED/attrs',
```
Comment (fgalan) [inline test/unit/ngsiv2/MQTT_receive_measures-test.js]: JSON_UNPROVISIONED?

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+    describe('When a POST measure arrives for the HTTP binding', function() {
+        var optionsMeasure = {
+            url: 'http://localhost:' + config.http.port + '/iot/json',
+            method: 'POST',
+            json: {
+                humidity: '32',
+                temperature: '87',
+                luminosity: 10,
+                pollution: 43.4,
+                configuration: {firmware: {verson: '1.1.0', hash: 'cf23df2207d99a74fbe169e3eba035e633b65d94'}},
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Little typo: verson -> version

**Comment by fgalan** [comment]
After a last round of very minor comments, I think the PR would be ready :)

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+            },
+            headers: {
+                'fiware-service': 'smartGondor',
+                'fiware-servicepath': '/gardens'
+            },
+            qs: {
+                i: 'MQTT_2',
+                k: '1234'
+            }
+        };
```
[the remark this answers is not in the document]
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed in 10969e0f085eab0d57d2bd42df675f732c2f7629

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 18 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed in 10969e0f085eab0d57d2bd42df675f732c2f7629

File: test/unit/ngsiv2/MQTT_receive_measures-test.js (hunk 17 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/MQTT_receive_measures-test.js]: Fixed in 10969e0f085eab0d57d2bd42df675f732c2f7629

File: test/unit/ngsiv2/MQTT_receive_measures-test.js (hunk 16 above)
Comment (dcalvoalonso) [reply test/unit/ngsiv2/MQTT_receive_measures-test.js]: Fixed in 10969e0f085eab0d57d2bd42df675f732c2f7629

File: package.json (hunk 15 above)
Comment (dcalvoalonso) [reply package.json]: You are right, it was just for that test. I forgot to delete it. :(

Comment (dcalvoalonso) [reply package.json]: Fixed in 10969e0f085eab0d57d2bd42df675f732c2f7629

**Comment by fgalan** [comment]
LGTM. Good work!

However, before merging we should:

1. First merge https://github.com/telefonicaid/iotagent-node-lib/pull/556 when it gets ready
2. Change `packages.json` in this PR to point back againt to master branch in the library dependency
3. Merge this PR

File: package.json [hunk 20]
```diff
@@ -47,6 +47,7 @@
   },
   "keywords": [],
   "dependencies": {
+    "express": "^4.11.2",
```
Comment (fgalan) [inline package.json]: A bit surprissed... was express removed in a previous commit?

Comment (dcalvoalonso) [reply package.json]: I did this error when merging changes of master branch... :(

Comment (dcalvoalonso) [reply package.json]: NTC

**Comment by fgalan** [comment]
LGTM to the changes in https://github.com/telefonicaid/iotagent-json/compare/d158ef4...8f6ff00

(When new commits after LGTM, then it is a good idea to "renew" the LGTM reviewing the new changes)

**Comment by fgalan** [comment]
LGTM to changes https://github.com/telefonicaid/iotagent-json/compare/8f6ff00...32effbb
